x/programs/rust/examples/token/src/lib.rs

```diff
@@ -25,27 +38,38 @@ pub fn mint_to(state: State, recipient: Address, amount: i64) -> bool {
         .is_ok()
 }
 
-/// Transfers amount coins from the sender to the recipient. Returns whether successful.
+/// @notice Transfers coins from the sender to the recipient
+/// @param state The current program state
+/// @param sender The address sending tokens
+/// @param recipient The address receiving tokens
+/// @param amount The amount of tokens to transfer
+/// @return Whether the transfer succeeded
 #[public]
 pub fn transfer(state: State, sender: Address, recipient: Address, amount: i64) -> bool {
     // require sender != recipient
-    if sender == recipient {
-        return false;
-    }
+    assert_ne!(sender, recipient);
+
     // ensure the sender has adequate balance
     let sender_balance: i64 = state.get_map_value("balances", &sender).unwrap_or(0);
-    if amount < 0 || sender_balance < amount {
-        return false;
-    }
+    assert!(amount >= 0 && sender_balance >= amount);
+
     let recipient_balance: i64 = state.get_map_value("balances", &recipient).unwrap_or(0);
+
+    let sender_delta = sender_balance.checked_sub(amount);
+    let recipient_delta = recipient_balance.checked_add(amount);
+    assert!(sender_delta.is_some() && recipient_delta.is_some());
```

Comment:
same here I think we want to consider better error handling vs panic right?

Reply:
This isn't idiomatic. Why waste computation if `sender_delta` is `None`? Prefer using `expect` if you want to panic.

Reply:
Alernatively, you could do something like this:
```rust
let (Some(sender_delta), Some(recipient_delta)) = (sender_delta, recipient_delta) else {
    panic!("transfer failed");
};
```
but it's still a wasteful computation if the first fails.